### PR TITLE
RavenDB-14753 - Counters are not being deleted from storage when we are changing casing but they are deleted from document metadata

### PR DESCRIPTION
**Pull request by shiranshalom**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-14753

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/DocumentIdWorker.cs [hunk 1]
```diff
@@ -181,6 +212,28 @@ private static ByteStringContext.InternalScope UnlikelyGetLowerUnicode(ByteStrin
 
         }
 
+        private static ByteStringContext.InternalScope UnlikelyGetLowerUnicode(ByteStringContext byteStringContext, string str, out Slice loweredKey)
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentIdWorker.cs:215]: This is duplicate code, no?

File: src/Sparrow/Json/LazyStringValue.cs [hunk 2]
```diff
@@ -84,6 +84,58 @@ public int Length
             }
         }
 
+        public bool EqualsOrdinalIgnoreCase(LazyStringValue other)
+        {
+            return CompareToOrdinalIgnoreCase(this.Buffer, Size, other.Buffer, other.Size) == 0;
+        }
+
+
+        public static int CompareToOrdinalIgnoreCase(byte* strA, int strALen, byte* strB, int strBLen)
```
Comment (ayende) [inline src/Sparrow/Json/LazyStringValue.cs:93]: Duplicate, no?

Comment (shiranshalom) [reply src/Sparrow/Json/LazyStringValue.cs:93]: No, I took these methods from the changes made at `LazyStringValue.cs` [ don't exist at V4.2 ].

The problem occurred when we tried to get the counter's Index by counter name [ which suppose to be case insensitive ] - but we used Inline compare (case sensitive).

File: src/Raven.Server/Documents/DocumentIdWorker.cs (hunk 1 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/DocumentIdWorker.cs:215]: No, I took these methods from https://github.com/ravendb/ravendb/commit/dacfed6aab0c9b901011ea5537580414bec93d48 which don't exist at V4.2

**Comment by arekpalinski** [comment]
@aviv86 we don't have the schema update here. I think it's acceptable in this case to not fix the existing records but I'd like to hear your opinion on that.

**Review by aviv86 — APPROVED** [review]

**Review by aviv86 — COMMENTED** [review]
in 5.0 we also keep original-casing of the counter names so that document's metadata keeps them in their original casing. 
this might be breaking a few tests?

**Comment by aviv86** [comment]
@arekpalinski Agreed, I also think it's acceptable in this case to not lower-case counter names in existing records. 
We have the schema upgrade in 5.0 which does that

**Comment by arekpalinski** [comment]
> @arekpalinski Agreed, I also think it's acceptable in this case to not lower-case counter names in existing records. We have the schema upgrade in 5.0 which does that

So does it make sense to apply this fix to 4.2?

**Comment by ppekrol** [comment]
Please check failing tests.

**Comment by aviv86** [comment]
> > @arekpalinski Agreed, I also think it's acceptable in this case to not lower-case counter names in existing records. We have the schema upgrade in 5.0 which does that
> 
> So does it make sense to apply this fix to 4.2?

Thinking about it again, this might be a problem 

If a 4.2 user has existing counters with non lowered cased names, and after this fix the new counters he adds are now all lower-cased, then the lookup with `ignoreCase = true` [here](https://github.com/ravendb/ravendb/pull/13112/files#diff-fd38cbbe3f3841d14dc8b936f33bfb379549421c85438280fabea880ee240a5dR1367 ) might not work

we can't mix lowered-cased and non-lowered in the same blittable and expect the case-insensitive search to work, because blittable properties are sorted in a case-sensitive manner and we're doing a binary search on them..

So I think we should either do a schema upgrade here, or apply this to 5.x only (counter names are already lowered cased there, so just the lookup with `ignoreCase = true` fix)

**Comment by arekpalinski** [comment]
Okay so this is problematic change for 4.2. I'd prefer to avoid doing schema upgrade since that's not so critical. Especially that applying this change to 5.x might be way easier. @shiranshalom is going to check that. If that's the case then I'm for applying it only for 5.2

**Comment by shiranshalom** [comment]
We are already handling that properly in 5.2 -  you can see a detailed explanation in https://issues.hibernatingrhinos.com/issue/RavenDB-14753#focus=Comments-67-140536.0-0

**Pull request closed without merging**